### ISSUE TITLE
Pyspark error: Could not initialize class io.grpc.netty.NettyChannelBuilder

### ISSUE THREAD

**Issue by esadler-hbo** [pinecone-io/spark-pinecone#5]
This time I am actually stumped. I believe there is a problem with an underlying java dependency. I imagine this related to the README mention of shading packages. Could really use some advice.

I did notice that some packages were evicted, but I am not sure how to fix that or if it really matters:

### Evicted Modules
```
	:: evicted modules:
	com.google.protobuf#protobuf-java;3.19.2 by [com.google.protobuf#protobuf-java;3.19.4] in [default]
	com.google.api.grpc#proto-google-common-protos;2.0.1 by [com.google.api.grpc#proto-google-common-protos;2.8.3] in [default]
	com.google.protobuf#protobuf-java;3.13.0 by [com.google.protobuf#protobuf-java;3.19.2] in [default]
	com.google.errorprone#error_prone_annotations;2.7.1 by [com.google.errorprone#error_prone_annotations;2.10.0] in [default]
	io.netty#netty-tcnative-classes;2.0.46.Final by [io.netty#netty-tcnative-classes;2.0.50.Final] in [default]
```

### Code
```python
from pyspark import SparkConf
from pyspark.sql import SparkSession, DataFrame
from pyspark.sql.types import StructType, StructField, ArrayType, FloatType, StringType
import pandas as pd
import numpy as np

import json

spark = SparkSession \
    .builder \
    .appName("Python Spark SQL basic example") \
    .config('spark.jars.packages', 'io.pinecone:spark-pinecone_2.12:0.1.1') \
    .getOrCreate()


metadata = json.dumps({"type": "test", "tags": ["a", "b", "c"], "size": 10000})
df = pd.DataFrame({
    "id": ["1"],
    "values": [np.random.normal(size=2048).astype(np.float32).tolist()],
    "namespace": ["default"],
    "metadata": [""]
})
schema = StructType([
    StructField("id", StringType(), False),
    StructField("values", ArrayType(FloatType()), False),
    StructField("namespace", StringType(), True),
    StructField("metadata", StringType(), True),
])


df = spark.createDataFrame(df, schema=schema)

(df.write
    .option("pinecone.apiKey", PINECONE_API_TOKEN)
    .option("pinecone.environment", ENV)
    .option("pinecone.projectName", PROJECT_NAME)
    .option("pinecone.indexName", INDEX_NAME)
    .format("io.pinecone.spark.pinecone.Pinecone")
    .mode("append")
    .save())
```

### Error
```
23/03/27 13:44:30 ERROR Executor: Exception in task 4.0 in stage 0.0 (TID 4)
java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder
	at io.pinecone.PineconeConnection.buildChannel(PineconeConnection.java:89)
	at io.pinecone.PineconeConnection.<init>(PineconeConnection.java:44)
	at io.pinecone.PineconeClient.connect(PineconeClient.java:37)
	at io.pinecone.spark.pinecone.PineconeDataWriter.<init>(PineconeDataWriter.scala:27)
	at io.pinecone.spark.pinecone.PineconeDataWriterFactory.createWriter(PineconeDataWriterFactory.scala:10)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:430)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:381)
```

**Comment by esadler-hbo** [comment]
I was able to solve the issue by rebuilding the library locally and referencing the jar. These were my steps for anyone else who wants to use spark with pinecone. 

```sh
git clone git@github.com:pinecone-io/spark-pinecone.git
cd spark-pinecone
sbt assembly
```

```python
spark = SparkSession \
    .builder \
    .appName("Python Spark SQL basic example") \
    .config('spark.jars', '/path/to/spark-pinecone/target/scala-2.12/spark-pinecone-0.1.1.jar') \
    .getOrCreate()
```

**Comment by rajat08** [comment]
> 

Thanks for notifying us of this @esadler-hbo! 
Depending on where you are running your Spark clusters there may be some grpc conflicts with the underlying Java client. We mention shading because we use this to run our workloads in Databricks. The current assembly rules are designed specifically to work with Databricks but it may also work for you on other Spark envs. We also publish a shaded version of the library in maven for this reason.

I will keep this in mind when we are updating our Java client next, thanks again

**Comment by olgaakrekhovetska** [comment]
hey @rajat08 !
I am running ingestion from Databricks notebook and getting the same issue
`[WRITING_JOB_ABORTED] Writing job aborted
Caused by: Job aborted due to stage failure: Could not initialize class io.grpc.netty.NettyChannelBuilder
Caused by: NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder
	at org.apache.spark.sql.errors.QueryExecutionErrors$.writingJobAbortedError(QueryExecutionErrors.scala:996)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:419)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2$(WriteToDataSourceV2Exec.scala:363)
	at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.writeWithV2(WriteToDataSourceV2Exec.scala:249)
	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run(WriteToDataSourceV2Exec.scala:342)
	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run$(WriteToDataSourceV2Exec.scala:341)
	at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.run(WriteToDataSourceV2Exec.scala:249)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$$nestedInanonfun$eagerlyExecuteCommands$1$1.$anonfun$applyOrElse$2(QueryExecution.scala:229)`

What version should be used in order get the appropriate libraries shaded?
Both from https://mvnrepository.com/artifact/io.pinecone/spark-pinecone don't work.

**Comment by rajat08** [comment]
> hey @rajat08 ! I am running ingestion from Databricks notebook and getting the same issue `[WRITING_JOB_ABORTED] Writing job aborted Caused by: Job aborted due to stage failure: Could not initialize class io.grpc.netty.NettyChannelBuilder Caused by: NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder at org.apache.spark.sql.errors.QueryExecutionErrors$.writingJobAbortedError(QueryExecutionErrors.scala:996) at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:419) at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2$(WriteToDataSourceV2Exec.scala:363) at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.writeWithV2(WriteToDataSourceV2Exec.scala:249) at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run(WriteToDataSourceV2Exec.scala:342) at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run$(WriteToDataSourceV2Exec.scala:341) at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.run(WriteToDataSourceV2Exec.scala:249) at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43) at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43) at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49) at org.apache.spark.sql.execution.QueryExecution$$anonfun$$nestedInanonfun$eagerlyExecuteCommands$1$1.$anonfun$applyOrElse$2(QueryExecution.scala:229)`
> 
> What version should be used in order get the appropriate libraries shaded? Both from https://mvnrepository.com/artifact/io.pinecone/spark-pinecone don't work.

@olgaakrekhovetska , can you try to get the shaded version with the assembly classifier ? 
```
<dependency>
  <groupId>io.pinecone</groupId>
  <artifactId>spark-pinecone_2.12</artifactId>
  <version>0.1.1</version>
  <classifier>assembly</classifier>
</dependency>
``` 

If that does not work, I can add an artifact to the repo and document it

**Comment by olgaakrekhovetska** [comment]
@rajat08 It doesn't work for sbt-based projects.
The same for Databricks notebook, cause there you add library to the cluster by downloading it directly from Maven repository and then you get error in runtime.
As a workaround, you can clone https://github.com/pinecone-io/spark-pinecone, assemble jar and add it to Databricks cluster from you local machine, in such case all goes smoothly. 

Wouldn't you mind to upload a new version to Maven?

**Comment by rajat08** [comment]
@olgaakrekhovetska , apologies that this flew under the radar. I wasn't aware it didn't work for sbt based projects, thanks for pointing it out. There is a planned release we hope to get to very soon

**Comment by olgaakrekhovetska** [comment]
Is there any ETA?

**Comment by rajat08** [comment]
> Is there any ETA?

Definitely in a few weeks, sorry I don't really have a concrete deadline. I would reach out to support, which would give us a channel to escalate if it's urgent for you.

**Comment by rohanshah18** [comment]
Hi @olgaakrekhovetska, thanks for reporting the issue. What versions of java, scala, and spark is causing you the NettyChannelBuilder error? With java 8, scala 2.12.5, spark-pinecone 2.12:0.1.1, and spark 3.2.0, I am able to run the sbt project ([from BasicIntegrationSpec.scala example](https://github.com/pinecone-io/spark-pinecone/blob/main/src/it/scala/io/pinecone/spark/pinecone/BasicIntegrationSpec.scala)) without having the need of assembled jar. But I'd like to replicate the same issue to understand what dependency version conflict is causing this issue and roll out a fix. 

I created a sbt project for the BasicIntegrationSpec.scala example I mentioned before and used the following as dependencies in sbt file, for your reference:

```
ThisBuild / scalaVersion := "2.12.15"
libraryDependencies ++= Seq(
  "org.scala-lang" % "scala-library" % scalaVersion.value,
  "io.pinecone" % "spark-pinecone_2.2" % "0.1.1",
  "org.apache.spark" %% "spark-core" % "3.2.0",
  "org.apache.spark" %% "spark-sql" % "3.2.0",
  "org.apache.spark" %% "spark-catalyst" % "3.2.4" % Test
)
```

**Comment by olgaakrekhovetska** [comment]
Hey @rohanshah18 !

The issue is caused when fat jar is deployed and run in Databricks.

Here is how build.sbt look like
`
//Spark version has to be >=3.2.0 to be compatible with Spark Pinecone connector
val sparkVersion         = "3.4.1"
val scala212 = "2.12.10"
val sparkPineconeVersion = "0.1.1"
libraryDependencies ++= Seq(  
  "org.apache.spark" %% "spark-core" % sparkVersion % Provided,
  "org.apache.spark" %% "spark-sql"  % sparkVersion % Provided,
  "io.pinecone" %% "spark-pinecone" % sparkPineconeVersion,
)
`
I am getting errors by running fat jar in different Databricks Runtimes:

- 13.3 LTS that stands for Spark 3.4.1 , Scala 2.12
- 11.3 LTS Spark 3.3.0, Scala 2.12(in order to use this one you will need to change Spark version in build.sbt)
The error message is all the same:
`Caused by: IllegalArgumentException: cannot find a NameResolver for <index_name>-<project_id>.svc.us-east4-gcp.pinecone.io`, `index_name` and `project_id` are populated with real names that are relevant to Pinecone index in our setup.

I don't see Databricks Runtime that goes with Spark 3.2.0, that is why picked those that have Spark version >=3.2.0 pre-installed.

**Comment by rohanshah18** [comment]
@olgaakrekhovetska Thanks! I'm able to replicate the `NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` error on databricks platform for Spark 3.4.1 and Scala 2.12. I'll now start to investigate and fix this error, and eventually publish a new jar (so one can import the pinecone spark connector by downloading the jar from maven central). 

Although it is interesting that the sbt or mvn projects built on my machine without the fat jar, are able to run successfully on Spark 3.4.1 and Scala 2.12.

Once I'm done fixing this issue, I'll try to replicate your error of `Caused by: IllegalArgumentException: cannot find a NameResolver for <index_name>-<project_id>.svc.us-east4-gcp.pinecone.io` on my local and fix it as well.

**Comment by rohanshah18** [comment]
@esadler-hbo @olgaakrekhovetska I have published the new version of spark-pinecone connector v0.1.4 and tested it on databricks (DBR 14.1: Spark 3.5.0 and Scala 2.12) with spark-pinecone assembly jar successfully. You can find the assembly jar for Spark 3.5.0  and Scala 2.12 here:
s3://pinecone-jars/spark-pinecone-uberjar.jar

I have also reached out to databricks on why the regular jar is still complaining about ClassNotFound error since I'm able to run it successfully in a sbt, gradle, or maven project. So they are investigating this issue for us. In the meantime, hoping this unblocks you and feel free to reach out if you have any issues.

**Comment by rohanshah18** [comment]
Closing this issue since I haven't heard back and hoping this is resolved.

**Comment by zbstof** [comment]
@rohanshah18 , I believe this bug should be re-opened.

We're using Databricks clusters to run the code.
The `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` problem does go away when attaching to the cluster the uberjar (tested both `s3://pinecone-jars/spark-pinecone-uberjar.jar`, `s3://pinecone-jars/0.2.1/spark-pinecone-uberjar.jar`).
But when building sbt application jar by incorporating `spark-pinecone` jars from maven repository:

> `libraryDependencies += "io.pinecone" %% "spark-pinecone" % "0.1.4"`

... we still get the same exception. Also reproduced for version `0.2.1`

> I have also reached out to databricks on why the regular jar is still complaining about ClassNotFound error since I'm able to run it successfully in a sbt, gradle, or maven project. So they are investigating this issue for us.

Have Databricks reached out to you with any update?

**Comment by rohanshah18** [comment]
@zbstof Thanks for reaching out with your issue. Want to make sure we are on the same page, so please let me know if any of the following is not true:
1. I'm glad that you're able to run spark-pinecone on Databricks platform with s3 jar successfully.
2. You're facing `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` error when running a sbt style project on your local with the thin jar:
`libraryDependencies += "io.pinecone" %% "spark-pinecone" % "0.1.4"` or with 0.2.1.

One thing I already notice is `spark-pinecone` in sbt dep. should have scala version appended, so it should be spark-pinecone_2.12 if using scala 2.12 or 2.13 if using scala 2.13. Below is the example:
`"io.pinecone" % "spark-pinecone_2.12" % "0.2.1"`
but you can find the dependency here: https://central.sonatype.com/search?q=spark-pinecone.

I can confirm that I am successfully able to run sbt-style project on my local with `spark-pinecone-0.2.1` jar and I made sure to delete the pre-downloaded jars from maven central in `/.ivy2`, so I get fresh copies from the maven central. Please find below the screenshot as a proof. I have highlighted the deps for you on the left and the right and the main file shows the code I ran to upsert vectors:
![image](https://github.com/pinecone-io/spark-pinecone/assets/9093060/ad73a6a2-c454-47e3-89a0-fa8ed12a8a67)

I'll strongly recommend you to check your versions compatible with spark-3.5.0:
java: 8
scala: 2.12 or 2.13
spark: 3.5.0
gRPC version: 1.58.0 (which internally uses netty version: 4.1.93.Final)

Another issue can be transitive or direct dependency clash on your project if you are using extra libraries other than the ones I have in the sample test project. I usually like to print dependency tree first, if I were to resolve this issue.

As far as databricks is concerned, they said there's no way to print dependency tree in the notebooks, so I'm trying out alternate approaches to reproduce this issue locally.

**Comment by olgaakrekhovetska** [comment]
@rohanshah18 `%%` syntax in `built.sbt` will take care of appending Scala version to the artifact name, so we should be good here.

The issue itself can be reproduced by spinning up Databricks job leveraging [DBR 14.1](https://docs.databricks.com/en/release-notes/runtime/14.1.html) and uploading custom jar assembled by sbt. This jar should have `build.sbt` that has spark-pinecone connector dep stated in `libraryDependencies `.
I have also checked DBR 14.1 pre-installed libraries and it looks like netty is version of `4.1.96.Final`.

**Comment by rohanshah18** [comment]
@olgaakrekhovetska Yes, we are still facing the dependency issue on DBR platform using the thin jar and that's why we have released the newer version of fat jar on s3 bucket. Is there a requirement on your side that's stopping you from using fat jar temporarily? I have received a long list of dependencies from Databricks to replicate the runtime environment locally so I can reproduce this issue (on my local).

**Comment by zbstof** [comment]
I've noticed that fat jar shades guava and protobuf libraries, but plain jar does not.
This is a problem because as `spark-pinecone`'s dependecy `io.grpc:grpc-netty:1.58.0` depends in turn on `com.google.guava:guava:32.0.1-android`, the version installed on DBR 14.1 is `com.google.guava:guava:15.0`, and running snippet on Databricks:
```
import io.grpc.netty.NettyChannelBuilder
NettyChannelBuilder.forTarget("some-endpoint")
```
throws an error 
```
NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;CLjava/lang/Object;)V
```
which translates to signature `void checkArgument(boolean, String, char, Object)` which does not exist on guava:15.0

What is weird is that I cannot reproduce this error, but I'm pretty sure that `NoSuchMethodError` causes an exception in intilaization logic of `NettyChannelBuilder`, which causes `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder`.
I've lost complete stacktrace, but most likely this happens through chain `io.grpc.internal.GrpcUtil` -> `io.grpc.Metadata.Key.of` -> `io.grpc.Metadata.AsciiKey#<init>` -> `io.grpc.Metadata.Key#validateName`

I've seem to have fixed `NoClassDefFoundError` by shading `guava` in sbt project:
```
ThisBuild / assembly / assemblyShadeRules := Seq(
  ShadeRule
    .rename("com.google.common.**" -> "shaded.guava.@1")
    .inAll,
)
```

But I get another exception now when trying to run application:
```
IllegalArgumentException: cannot find a NameResolver for interaction-doc-54ca5b4.svc.us-east4-gcp.pinecone.io
	at io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:742)
	at io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:751)
	at io.grpc.internal.ManagedChannelImpl.<init>(ManagedChannelImpl.java:631)
	at io.grpc.internal.ManagedChannelImplBuilder.build(ManagedChannelImplBuilder.java:631)
	at io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:297)
	at io.pinecone.PineconeConnection.buildChannel(PineconeConnection.java:108)
	at io.pinecone.PineconeConnection.<init>(PineconeConnection.java:46)
	at io.pinecone.PineconeClient.connect(PineconeClient.java:39)
	at io.pinecone.spark.pinecone.PineconeDataWriter.<init>(PineconeDataWriter.scala:27)
	at io.pinecone.spark.pinecone.PineconeDataWriterFactory.createWriter(PineconeDataWriterFactory.scala:10)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run(WriteToDataSourceV2Exec.scala:493)
```
EDIT: also shaded `"com.google.protobuf.**" -> "shaded.protobuf.@1"`, same error

**Comment by zbstof** [comment]
So maybe this issue should be re-opened, what do you think? Because I think the solution to `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` would be to package the jar together with shaded guava. Or should I create another issue for `NameResolver` subissue?

**Comment by rohanshah18** [comment]
@zbstof Please feel free to open a support ticket for NameResolver error. As far as NettyChannelBuilder error is concerned, you're right. I'm working to fix this issue but just to confirm, you were running sbt project on your local right? If so, then the thin jar should be enough to run the upserts successfully.

For NameResolver, what operations were you running? Did you try to run describeIndexStats right after index creation?

**Comment by rohanshah18** [comment]
@zbstof Please feel free to open a support ticket for NameResolver error. Also what operations were you running? Did you try to run describeIndexStats right after index creation?

For `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` error, I'm unable to reproduce this locally. I tried adding guava 15 to sbt project that also consumes spark-pinecone client. The result is grpc-netty (guava 32) will evict guava 15 and works successfully. But the existing spark-pinecone sdk's thin jar has been working on sbt, mvn, and gradle project locally.

I'm really interested in knowing if you were able to reproduce this error on sbt or maven or gradle project locally since you mentioned:

> I've seem to have fixed NoClassDefFoundError by shading guava in sbt project

So how did you reproduce this locally? I can reproduce this on databricks and see the driver logs, so I 100% agree that guava 15 is clashing with grpc-netty's guava 32.

**Comment by rohanshah18** [comment]
@zbstof If I add guava 15 in sbt using `dependencyOverrides`, then I can reproduce this error locally since I'm forcing sbt to not use the highest version. So the only explanation I can think of is databricks is not resolving dependency conflict to use the highest version compared to what sbt does. I'll have this confirmed with the databricks team.

**Comment by rohanshah18** [comment]
@zbstof Assembly jar already works on databricks platform which has the following shading rules:
```
assembly / assemblyShadeRules := Seq(
      ShadeRule
        .rename("com.google.protobuf.**" -> "shaded.protobuf.@1")
        .inAll,
      ShadeRule
        .rename("com.google.common.**" -> "shaded.guava.@1")
        .inAll
    ),
```

Is `NameResolver` the only error you need help with? If so, then I'll be closing this ticket since a customer support ticket will help you guide better and faster.

**Comment by zbstof** [comment]
> I'm really interested in knowing if you were able to reproduce this error on sbt or maven or gradle project locally
No, I didn't try to reproduce locally. I've just tested on databricks

Huh. Looks like there's no way to fix version that is published to mvncentral.
We should ask Databricks to upgrade bundled guava version to 32

> Is NameResolver the only error you need help with? If so, then I'll be closing this ticket since a customer support ticket will help you guide better and faster.

Ok, I'll take it elsewhere. It's just only this NameResolver problem is an extension of `java.lang.NoClassDefFoundError: Could not initialize class io.grpc.netty.NettyChannelBuilder` error and the setup to reproduce is almost identical (just need to add shade rule as described above to get NameResolver error).

**Comment by rohanshah18** [comment]
@zbstof I did reach out to databricks and they said they can't upgrade guava version yet since Spark 3.5 relies on the older version (guava 14).  For NameResolver error, I am not sure yet if its an extension of NettyChannelBuilder error since I'm not able to reproduce it locally. But if you believe its an extension, please feel free to open a new github issue. I'd request you to provide me with details on what code you were running and your setup details, which will also be asked by the github issue template. 

I'll close this issue so we can isolate everything in one place for NameResolver error.
